instrumentation/opentelemetry-instrumentation-sqlalchemy/src/opentelemetry/instrumentation/sqlalchemy/engine.py

```diff
@@ -100,42 +100,37 @@ def _before_cur_exec(self, conn, cursor, statement, *args):
             self._operation_name(db_name, statement),
             kind=trace.SpanKind.CLIENT,
         )
-        self.current_thread_span = self.cursor_mapping[cursor] = span
         with trace.use_span(span, end_on_exit=False):
             if span.is_recording():
                 span.set_attribute(SpanAttributes.DB_STATEMENT, statement)
                 span.set_attribute(SpanAttributes.DB_SYSTEM, self.vendor)
                 for key, value in attrs.items():
                     span.set_attribute(key, value)
 
-    # pylint: disable=unused-argument
-    def _after_cur_exec(self, conn, cursor, statement, *args):
-        span = self.cursor_mapping.get(cursor, None)
-        if span is None:
-            return
+        context._span = span
 
-        span.end()
-        self._cleanup(cursor)
 
-    def _handle_error(self, context):
-        span = self.current_thread_span
-        if span is None:
-            return
+# pylint: disable=unused-argument
+def _after_cur_exec(conn, cursor, statement, params, context, executemany):
+    span = getattr(context, "_span", None)
+    if span is None:
+        return
 
-        try:
-            if span.is_recording():
-                span.set_status(
-                    Status(StatusCode.ERROR, str(context.original_exception),)
-                )
-        finally:
-            span.end()
-            self._cleanup(context.cursor)
-
-    def _cleanup(self, cursor):
-        try:
-            del self.cursor_mapping[cursor]
-        except KeyError:
-            pass
+    span.end()
+
+
+def _handle_error(context):
+    span = getattr(context.execution_context, "_span", None)
+    if span is None:
+        return
+
+    try:
+        if span.is_recording():
+            span.set_status(
+                Status(StatusCode.ERROR, str(context.original_exception),)
+            )
```

Comment:
Do we expect the `str()` call to fail in some situations? if not, why put this under a try block?

Reply:
Went back through history and it seems to have always been that way - but I can't imagine why. Even if it's `None` it will just become the `"None"`. I've moved it out of the `try catch` block but want to check what should be passed into `Status` if no exception is available.

Reply:
`original_exception` is always present: https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.ExceptionContext.original_exception so should be safe :+1: